### DISCUSSION TITLE
Possible to pass through exact text of disallowed types?

### DISCUSSION THREAD

**Discussion by ryana** [remarkjs/remark#637] [Q&A] [answered]
I am building a chat product, and I want my users to be able to us a subset of markdown, so I've done this:

```
import ReactMarkdown from 'react-markdown';
import {autoLinkMarkdown} from "components/common/AutoLink";
import gfm from 'remark-gfm';

const CustomMarkdown = (props) => {
  const { source, allowedTypes } = props;
  const allowedTypesWithDefaults = allowedTypes || ["link", "break", "paragraph", "root", "text", "list", "listItem", "strong", "emphasis"];

  return (
    <ReactMarkdown
      escapeHtml={false}
      skipHtml={true}
      unwrapDisallowed={true}
      source={ autoLinkMarkdown( { text:source } ) }
      allowedTypes={allowedTypesWithDefaults}
      renderers={ { link: props => <a href={props.href} rel="nofollow noopener" target="_blank" >{props.children}</a>} }
      />
  )
}

export default CustomMarkdown;
```

(the `autoLinkMarkdown` is turning @-mentions into markdown links, but not otherwise modifying the source)

The problem is that if they type this:

```
Here is a list of things that I'm starting with a hash for reasons beyond me:

# My first item
# My second item
# My third item
```

Then the output looks like this:

```
Here is a list of things that I'm starting with a hash for reasons beyond me:

My first item
My second item
My third item
```

In the output, the hashmarks are missing. I assume it has something to do with the tokenizer swallowing characters for the disallowed `heading` node type.

Similarly, if someone decides to write an HTML comment (also disallowed):
```
Hey guys <!-- how you doing -->
```

Then all they see is:

```
Hey guys
```

I'd really like an experience where everything they type is visible in the output, while giving me the flexibility to do basic markdown and autolinking.

I saw the option [here](https://github.com/remarkjs/react-markdown/issues/86) to pass in my own renderers, and I'm guessing I could just write a pass through renderer for every other node listed [here](https://github.com/remarkjs/react-markdown#appendix-b-node-types) but that feels wonky so I wanted to ask before I went and did that.

Is there some really cool option staring me right in the fact that'll already do this for me? :)

Thx,

**Comment by wooorm** [comment, marked as answer]
The {dis}allowed features work after parsing. They’re more a way to sanitize the output. Not a way to change the input.

> I'd really like an experience where everything they type is visible in the output, while giving me the flexibility to do basic markdown and autolinking.

That wouldn’t be markdown. The goal of react-markdown and the underlying tools is to match exactly how CommonMark (and optionally GH) parse markdown, not to make it possible to support markdown-like formats.
Markdown is a really complex little language. There are some [low level hooks](https://github.com/remarkjs/remark/issues/552) to do something close to what you’re going for, but I wouldn’t recommend it.

> Similarly, if someone decides to write an HTML comment (also disallowed)

You have `skipHtml` turned on.

**Reply by wooorm** [reply]
Maybe you’ll be best helped with a couple of regexes [like so](https://github.com/developit/snarkdown)

**Reply by ryana** [reply]
> That wouldn’t be markdown. The goal of react-markdown and the underlying tools is to match exactly how CommonMark (and optionally GH) parse markdown, not to make it possible to support markdown-like formats.

This is a super helpful comment. I hadn't thought of it that way. I had looked at a bunch of markdown-esque-lite libraries including snarkdown, but I settled on react-markdown because it was fully featured (including things like lists) and because I liked the fact it saved me from having to use `dangerouslySetInnerHTML`, but your comment is going to make me revisit that.

Thanks a lot for taking the time to respond!
